### ISSUE TITLE
How does the upstream Crosswire syncing work?

### ISSUE THREAD

**Issue by lancejpollard** [getbible/v2#11]
I checked out https://github.com/getbible/v2_builder, and love your API. I would like to collect and publish bibles (and other ancient and modern texts) and potentially provide an API for them, but I get your concern that texts always have errors and need to be fixed, so publishing a GitHub repo is probably bad practice (unless you synced all updates to the repo??) because forks will quickly get out of sync.

However, I don't see any description of how the syncing works with Crosswire. First, what is Crosswire, and how does getbible.com relate to it? Second, why are they considered the source of truth? What other orgs can act of sources of bible truth, and why didn't you choose those but instead chose Crosswire? I'm curious how all the decisions were made :D. The only context I have is from your v2_builder paragraph, which doesn't go into enough detail :)

> Well, because we at getBible would like to comply with the Crosswire conventions to remain in sync with their modules all the way downstream from our project. Those who do not honor these agreements are responsible for Scripture Text (Digital) being incorrectly distributed with errors, as the digital versions of these Bibles may contain spelling mistakes, typos, or other errors like missing verses or chapters. So as they are discovered and fixed, Crosswire releases updates to their modules, and our commitment to the above agreed conventions ensure that these updates flow downstream. Should you have any further questions please do not hesitate to open an issue on this repository. Honestly, without these measures there is no getBible project... it is that serious.

Basically, what is your strategy, and how is it implemented (at a high level)? How are you syncing, and how are you getting notified of updates whereby you then resync? That sort of stuff. And yeah, why Crosswire was chosen as the source of truth?

**Comment by Llewellynvdm** [comment]
@lancejpollard my apologies for not getting back to you sooner.

Okay, essentially, CrossWire is not necessarily the source of truth. It's just an organization that has already done a lot of legwork to get permissions from all the different publishers to add those published Bibles to their infrastructure. Since their infrastructure allows publishers to manage their own Bible translations, it has become a highly convenient central community of people working on the same infrastructure. We had no ambition to start our own structure to see how GetBible.net deals with the actual task of keeping everything in sync.

You rightly refer to the [v2_builder](https://github.com/getbible/v2_builder) Repository. That repository has some scripts in its source (src) folder. These scripts first download the actual CrossWire modules from CrossWire. Once the modules are downloaded, Python and Bash are used to extract the actual text from the modules and convert them into a JSON representation of the scripture. We've done this in the most cautious way possible to avoid any changes to the actual text. However, in doing so, we're not transferring all the module functionality, because we're working with a JSON object. Currently, we only want the text, not all the other extra data that might exist in those modules.

That means if you want to see how the actual syncing is done, you need to look at this [v2_Builder/src](https://github.com/getbible/v2_builder/tree/master/src) folder. You need to read through the different scripts that are used to actually trigger the different tasks. There is also a .github/workflows folder, and in this folder, there is a build.yaml file. This build.yaml file gets triggered once a month. GitHub runs the build, so the downloading of the modules from CrossWire and the actual updating happens on GitHub through their workflows. This has been a very convenient and stable option for us so far.

That being said, if a build is done and a change in the text is detected, it will update a private repository, also hosted here on GitHub. It's not open to the public, and this private repository triggers a workflow that, once it's updated, pushes the changes to our various API S3 containers, making them available through the URLs.

The [v2](https://github.com/getbible/v2) Repository is public and provides the framework. All the keys to the repository are there, so you can easily detect and discover the necessary books and chapters of every translation. Using the public Version 2 repository as the baseline, you can access all of the actual scripture through our API.

This is our best effort so far. I'm continuously working on improving this. I've just recently started working on an additional interface that will run through Cloudflare and probably help improve the quality of our API.

I'm not sure if this entirely answers your question on how it works, because you need to read through the actual source code of the Builder to understand what it does. Essentially, that is the documentation.

The modules are maintained by the publishers on CrossWire. CrossWire maintains a few modules themselves, but there are several modules where the actual publishers of the scriptures maintain them through repositories, which CrossWire then builds into their modules.

Since going to all these publishers ourselves and getting all those permissions was something I wanted to avoid, that's why we use CrossWire. Our API also shares, within the scriptural linking, all the CrossWire information of each translation. That's the closest we've come to ensuring that the data is accurate, and that we meet all their expectations.

Yeah, I think the rest is simply the best we could do. I hope that answers your question.
